### ISSUE TITLE
hx-preserve does not work in oob-swapped content

### ISSUE THREAD

**Issue by irfinnew** [bigskysoftware/htmx#2922]
I'm having an issue where an element with `hx-preserve` set is not actually preserved if the element is updated through an OOB swap.

Proof of concept code on the main page:
```html
<div id="foo" hx-get="partial.html" hx-trigger="load"></div>

<p><button onclick="document.getElementById('bar').innerHTML = 'goodbye';">change</button></p>
<p><button hx-get="partial.html" hx-swap="none">update</button></p>

partial.html:
<div id="foo" hx-swap-oob="true">
	<p><span id="bar" hx-preserve="true">hello</span></p>
</div>
```

Pressing the `change` button changes `span#bar`'s content (expected), but `update` reverts this change (unexpected).

The same PoC _does_ work with a regular targeted non-OOB swap:

```html
<div id="foo" hx-get="partial.html" hx-trigger="load"></div>

<p><button onclick="document.getElementById('bar').innerHTML = 'goodbye';">change</button></p>
<p><button hx-get="partial.html" hx-target="#foo">update</button></p>

partial.html:
<div id="foo">
	<p><span id="bar" hx-preserve="true">hello</span></p>
</div>
```

I tested with htmx 1.8.6, 1.9.12, and 2.0.2, on both Chrome and Firefox. The OOB version didn't seem to work on any combination of them.

btw, thank you for making web development fun again :)

**Comment by Telroshan** [comment]
Hey, looking at the code, it seems that what's happening is:
- oob swapping is processed first
https://github.com/bigskysoftware/htmx/blob/b71af75f1af7d5d3ae562686625fe131f6923e5a/src/htmx.js#L1856
- it removes the element from the response fragment after swapping it
https://github.com/bigskysoftware/htmx/blob/b71af75f1af7d5d3ae562686625fe131f6923e5a/src/htmx.js#L1455
- only then, elements to preserve are processed (which happens _before_ the standard swap)
https://github.com/bigskysoftware/htmx/blob/b71af75f1af7d5d3ae562686625fe131f6923e5a/src/htmx.js#L1872-L1873

While it looks like we never supported `hx-preserve` along oob swap, I don't see why we wouldn't want to, and I'd consider this as a bug since the documentation doesn't mention any exception about oob swapping so one would expect it to work just like standard swapping.
So, if you'd like to investigate and submit a PR, please feel free to do so!

**Comment by MichaelWest22** [comment]
The code for processing hx-preserve is located a couple of blocks below the oob swap handling which is why oob swaps do not currently support this attribute.

Reading the code it seems it should be possible to move the handling line up above the oob swap code and then it would probably work fine.  However I do see one potential issue.  the feature works by updating the response fragment of html with the contents of the preserved elements so that when it swaps the data in it will have the same content as the current page contents.  But what happens when the page's element with matching id to the hx-preserve one has a different hx-swap-oob attribute value???  If it is missing the hx-swap-oob value it would disable the swap and that id would be placed with the main swap in the wrong location.  If the response has no hx-swap-oob value but the preserved id does have one for some reason it will turn the swap into oob.  And finially the type of hx-swap-oob could be set to a different value in the page causing the swap to target wrong.  So to fix this may have to always keep the hx-swap-oob value from the response somehow

**Comment by MichaelWest22** [comment]
looking a bit deeper I think if instead we add `handlePreservedElements(oobElement)` to the first line of the oobSwap() function and `restorePreservedElements()` just before the return of this function it will work and because the value of the hx-swap-oob has already been captured at this point the attribute state on the page will not have any negative impact.  Also have to change handlePreservedElements()'s parameter to also accept Element I think and add at least one test.

**Comment by MichaelWest22** [comment]
I just discovered something pretty amazing! Enabling oob swaps on hx-preserve gives htmx another super power that it didn't have before.  One of the emergent behaviors of hx-preserve is that the location in the DOM of the preserved element does not have to be the same after the swap. So for example you could have a video element in the bottom left corner of the page and if you return a new full or partial page that replaces the top right corner of the page with the same id and has hx-preserve set then it will pickup the content from the bottom left and replicate it up to the top right.  Also if you have the cannery version of chrome with the new experimental moveBefore feature turned on then the video element itself would move with no changes and keep playing after the relocation.  

Now once you unlock oob swaps this becomes even more powerful because you can now return a small bit of oob swap response in any request response and get it to place an element with an existing id and hx-preserve in any chosen location on the screen and it will make a copy of the element to the new location.  The provided oob item can also include a fall back copy of the content that will be used if the item is not found on the page or it can just be a empty tag if you know the content is on the page.  The only issue with this move feature is that unless they have experimental moveBefore enabled in the browser the original content will still remain so the item will then be located twice in the page after the oob swap completes.  It would be wonderful to have support for a `hx-relocate="true"` tag or make `hx-preserve="relocate"` work the same as how hx-preserve works now but also get it to remove the old content so that you won't end up with two copies.  Without this change instead you could do two oob swaps one after the other in the same response that moves the id to a new location and then delete the content in the old location.

For example This response will move the existing video element to instead be inside the top right div while also making sure it is removed from the bottom left div.
```html
<div hx-swap-oob="innerHTML:#toprightdiv"><div id="video" hx-preserve="true"></div></div>
<div hx-swap-oob="innerHTML:#bottomleftdiv"></div>
```

Also just tested locally with a few lines added to the preserve code to make hx-preserve="relocate" work and it moves the item fine with no delete required.

**Comment by MichaelWest22** [comment]
Added a PR to add the two lines required to enable obb swaps with hx-preserve.   Also found with more testing that I was wrong about needing any new relocate function it just already works fine as is and relocating works fine.  
Have a demo quick demo of oob swaps working here with the PR version of htmx.js: https://michaelwest22.github.io/htmx/index.html

**Comment by MichaelWest22** [comment]
@irfinnew fix now in v2.0.3 so please retest and close

**Comment by irfinnew** [comment]
Tested with v2.0.3, works for me now. Thanks all!
